src/sema/tests/mod.rs

```diff
@@ -128,3 +139,197 @@ fn test_statement_reachable() {
         assert_eq!(test_case.reachable(), expected);
     }
 }
+
+#[test]
+fn constant_overflow_checks() {
+    let file = r#"
+    contract test_contract {
+        function test_params(uint8 usesa, int8 sesa) public returns(uint8) {
+            return usesa;
+        }
+    
+        function test_add(int8 input) public returns (uint8) {
+            // value 133 does not fit into type int8.
+            int8 add_ovf = 127 + 6;
+    
+            // negative value -1 does not fit into type uint8. Cannot implicitly convert signed literal to unsigned type.
+            uint8 negative = 3 - 4;
+    
+            // value 133 does not fit into type int8.
+            int8 mixed = 126 + 7 + input;
+    
+            // negative value -1 does not fit into type uint8. Cannot implicitly convert signed literal to unsigned type.
+            return 1 - 2;
+        }
+    
+        function test_mul(int8 input) public {
+            // value 726 does not fit into type int8.
+            int8 mul_ovf = 127 * 6;
+    
+            // value 882 does not fit into type int8.
+            int8 mixed = 126 * 7 * input;
+        }
+    
+        function test_shift(int8 input) public {
+            // value 128 does not fit into type int8.
+            int8 mul_ovf = 1 << 7;
+    
+            // value 128 does not fit into type int8.
+            int8 mixed = (1 << 7) + input;
+        }
+    
+        function test_call() public {
+            // negative value -1 does not fit into type uint8. Cannot implicitly convert signed literal to unsigned type.
+            // value 129 does not fit into type int8.
+            test_params(1 - 2, 127 + 2);
+    
+            // negative value -1 does not fit into type uint8. Cannot implicitly convert signed literal to unsigned type.
+            // value 129 does not fit into type int8.
+            test_params({usesa: 1 - 2, sesa: 127 + 2});
+        }
+
+        function test_builtin (bytes input) public{
+
+            // value 4294967296 does not fit into type uint32.
+            int16 sesa = input.readInt16LE(4294967296);
+        }
+
+        function test_for_loop () public {
+            for (int8 i = 125 + 5; i < 300 ; i++) {
+            }
+        }
+
+        function composite(int8 a, bytes input) public{
+
+            uint8 sesa = 500- 400 + test_params(100+200, 0) + (200+101) + input.readUint8(4294967296);
+            int8 seas = (120 + 120) + a + (120 + 125);  
+            uint8 b = 255 - 255/5 ;
```

Comment:
This test verifies that, given a contract, you are generating the correct errors. It is not making sure that an input does not produce any error. Writing `uint8 b = 255 - 255/5` here has no effect for this test, because you do not check that it does not produce any diagnostic.

Reply:
In addition, `uint8 b = 300 - 255/5;` is a better test, as `255` fits in a 8-bit unsigned integer.

Reply:
I can implicitly test if it doesn't produce diagnostics by putting lines after it that will throw diagnostics. If there is one, the assert lines should fail as the diagnostics order will differ

Reply:
That makes sense. It is worth checking that the number of errors in this test is constant and equals a predetermined number. Otherwise, there may be new errors that we are unaware of.